framework/auth/views.py

```diff
@@ -202,92 +202,153 @@ def forgot_password_post(auth, **kwargs):
     return {}
 
 
-@collect_auth
-def auth_login(auth, **kwargs):
+def login_and_register_handler(auth, login=True, campaign=None, next_url=None, logout=None):
     """
-    This view serves as the entry point for OSF login and campaign login.
-    HTTP Method: GET
+    Non-view helper to handle `login` and `register` requests.
+
+    :param auth: the auth context
+    :param login: `True` if `GET /login`, `False` if `GET /register`
+    :param campaign: a target campaign defined in `auth.campaigns`
+    :param next_url: the service url for CAS login or redirect url for OSF
+    :param logout: used only for `claim_user_registered`
+    :return: data object that contains actions for `auth_register` and `auth_login`
+    :raises: http.BAD_REQUEST
+    """
+
+    # set target page to `dashboard` if no service url is specified
+    if not next_url:
+        next_url = web_url_for('dashboard', _absolute=True)
+
+    # Only allow redirects which are relative root or full domain. Disallows external redirects.
+    if not validate_service_url(next_url):
+        raise HTTPError(http.BAD_REQUEST)
+
+    data = {
+        'status_code': http.FOUND if login else http.OK,
+        'next_url': next_url,
+        'campaign': None,
+        'must_login_warning': False,
+    }
+
+    if campaign:
+        if validate_campaign(campaign):
+            # GET `/register` or '/login` with `campaign=institution`
+            # unlike other campaigns, institution login serves as an alternative for authentication
+            if campaign == 'institution':
+                data['status_code'] = http.FOUND
+                data['next_url'] = cas.get_login_url(next_url, campaign='institution')
+            # For all other non-institution campaigns
+            else:
+                # `GET /login?campaign=...`
+                if login:
+                    data['next_url'] = web_url_for('auth_register', campaign=campaign)
+                # `GET /register?campaign=...`
+                else:
+                    data['campaign'] = campaign
+                    data['next_url'] = campaigns.campaign_url_for(campaign)
+        else:
+            # invalid campaign
+            raise HTTPError(http.BAD_REQUEST)
 
-        GET '/login/' without any query parameter:
-            redirect to CAS login page with dashboard as target service
+    # if user is already logged in, overwrite the actions above and redirect to service url
+    if not logout and auth.logged_in:
+        data['status_code'] = http.FOUND
 
-        GET '/login/?logout=true
-            log user out and redirect to CAS login page with redirect_url or next_url as target service
+    # handle `claim_user_registered`
+    if logout and auth.logged_in:
+        data['status_code'] = 'auth_logout'
+        data['next_url'] = request.url
+        data['must_login_warning'] = True
```

Comment:
This is on the fringe of things that actually matter when it comes to changes in login/auth logic, but it might be worth verifying intended behavior of this message with product team. Not blocking